node/subsystem-util/src/lib.rs

```diff
@@ -203,6 +204,53 @@ specialize_requests! {
 	fn request_validation_code_hash(para_id: ParaId, assumption: OccupiedCoreAssumption)
 		-> Option<ValidationCodeHash>; ValidationCodeHash;
 	fn request_on_chain_votes() -> Option<ScrapedOnChainVotes>; FetchOnChainVotes;
+	fn request_session_executor_params(session_index: SessionIndex) -> Option<ExecutorParams>; SessionExecutorParams;
+}
+
+/// Requests executor parameters from the runtime effective at given relay-parent. First obtains
+/// session index at the relay-parent, relying on the fact that it should be cached by the runtime
+/// API caching layer even if the block itself has already been pruned. Then requests executor
+/// parameters by session index.
+/// Returns an error if failed to communicate to the runtime, or the parameters are not in the
+/// storage, which should never happen.
+/// Returns default execution parameters if the runtime doesn't yet support `SessionExecutorParams`
+/// API call.
+/// Otherwise, returns execution parameters returned by the runtime.
+pub async fn executor_params_at_relay_parent(
+	relay_parent: Hash,
+	sender: &mut impl overseer::SubsystemSender<RuntimeApiMessage>,
+) -> Result<ExecutorParams, Error> {
+	match request_session_index_for_child(relay_parent, sender).await.await {
+		Err(err) => {
+			// Failed to communicate with the runtime
+			Err(Error::Oneshot(err))
+		},
+		Ok(Err(err)) => {
+			// Runtime has failed to obtain a session index at the relay-parent; should never happen
+			Err(Error::RuntimeApi(err))
+		},
+		Ok(Ok(session_index)) => {
+			match request_session_executor_params(relay_parent, session_index, sender).await.await {
+				Err(err) => {
+					// Failed to communicate with the runtime
+					Err(Error::Oneshot(err))
+				},
+				Ok(Err(_)) => {
+					// Runtime doesn't yet support the api requested, should execute anyway
+					// with default set of parameters
+					Ok(ExecutorParams::default())
+				},
+				Ok(Ok(None)) => {
+					// Storage doesn't contain a parameter set for the given session; should
+					// never happen
+					Err(Error::RuntimeApi(RuntimeApiError::NotSupported {
+						runtime_api_name: "SessionExecutorParams",
+					}))
+				},
+				Ok(Ok(Some(executor_params))) => Ok(executor_params),
+			}
+		},
+	}
```

Comment:
It does makes sense to me. 

The only concern I have is that we're using a generic `Error` type here from subsystem-util. This poses a risk of treating the errors the same way on the caller side. However, I believe that `Oneshot` errors should be propagated, whereas `RuntimeApi` errors should be only logged at debug level. We're using `fatality` crate for that separation in many places. Example: https://github.com/paritytech/polkadot/blob/95468905f7f03289c353c3fe122b6bc96445aa5b/node/network/dispute-distribution/src/receiver/error.rs#L30

We could maybe use this Error type: https://github.com/paritytech/polkadot/blob/b7347e93527249899630bc8071b4e28a2003ce4a/node/subsystem-util/src/runtime/error.rs instead?

Reply:
I'm still not sure how to treat those errors. Pls have a look at [an example of usage](https://github.com/paritytech/polkadot/pull/6161/files#diff-c45578e383c1c7a84866f94a50b6f6725cc2e5d371df1bb8d5078f7f37d26545R540-R556).

The question is which is the best strategy when things are going really wrong. For example, if we're getting a runtime API error when trying to acquire a session index at relay parent. Right now I'm propagating that error to a higher level code which rejects the PVF being pre-checked. Probably it would be better to panic in that case because a failure to get the session index is kind of undefined behavior. But I'm not 100% sure which option is more correct.

Reply:
Agreed it's not a blocker for this PR, will be addressed in #6349